### PR TITLE
Medical Cadaver Crates

### PR DESCRIPTION
**Pull request by lizardqueenlexi**
## About The Pull Request

Adds a set of Medical Cadaver Crates, orderable through cargo, each of which contains one fully-intact and preserved corpse of a selected species! Any roundstart species can be ordered, with the exception of plasmamen - because I didn't have a good way to have them not burst into flames immediately. Also note that ethereal cadavers do not contain their crystal cores, for ethical reasons (NanoTrasen isn't about to ship you someone that will come back to life!). Each crate costs 1000 credits, and they can also be ordered through the medical orders console!

order
medorder

Each medical cadaver is shipped in a special "stasis coffin" - a medical coffin that preserves the organs of whatever is sealed inside, but stops working permanently once opened. No, it's not just an over-engineered freezer. Shhhh.

coffin

Finally, this PR contains a very rare easter egg... Don't look at the code if you don't want to spoil it, although maintainers probably should look at the code to tell me if it's actually a bad idea.

## Why It's Good For The Game

While roundstart morgue corpses of different species are nice, they're very limited. Sometimes, a situation will arrive where Medical would really like a corpse of a specific species, and there's just none on hand and no ethical way to make another one! This solves that problem. It's also a cargo-orderable source of organs, body parts, and species features, for if you really want to get up to some Frankensteinery.

Finally, heretics might enjoy it as an alternate means of getting your hands on some corpses, although ordering yourself a corpse through cargo is obviously a bit suspicious. Still, how are they gonna tell you no if you buy it with your own money, huh??

## Changelog

:cl:
add: Added orderable medical cadavers as cargo packs. Added an extremely rare easter egg that may happen when you order a medical cadaver.
/:cl:

### CODE REVIEW

File: orbstation/modules/cargo/cadavers.dm [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	desc = "A preserved medical cadaver, for dissections, spare organs, or experiments. This one is an ethereal. For legal reason, the body's crystal core has been removed."
+	contains = list(/mob/living/carbon/human/species/ethereal)
+
+// Remove the crystal core, both so the ethereal doesn't regenerate and so the crew can't bulk-order crystal cores
+/datum/supply_pack/medical/cadaver/ethereal/generate()
+	. = ..()
+	var/mob/living/carbon/human/H = locate() in .
+	for (var/part in H.internal_organs)
+		if(istype(part, /obj/item/organ/internal/heart))
+			qdel(part)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:46]: Does this kill the vampire?

File: orbstation/modules/cargo/cadavers.dm [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+/obj/structure/closet/crate/coffin/stasis/closet_update_overlays(list/new_overlays)
+	. = new_overlays
+	if(!opened && functional)
+		. += "stasisactive"
+	if(manifest)
+		. += "manifest"
+
+/obj/structure/closet/crate/coffin/stasis/open(mob/living/user, force = FALSE)
+	if(functional)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:109]: It's a style choice but I would prefer
```
if (!functional)
   return ..()

...everything else but now not indented quite so far...
return ..()
```

File: orbstation/modules/cargo/cadavers.dm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		functional = FALSE
+		for(var/atom/movable/AM in src)
+			if(istype(AM, /mob/living/carbon)) //unfreeze the organs of any contained carbon
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:112]: and similarly here
```
for(var/atom/movable/AM in src)
   if(!iscarbon(AM))
      continue
   ...rest of loop logic
```

File: orbstation/modules/cargo/cadavers.dm [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	. = new_overlays
+	if(!opened && functional)
+		. += "stasisactive"
+	if(manifest)
+		. += "manifest"
+
+/obj/structure/closet/crate/coffin/stasis/open(mob/living/user, force = FALSE)
+	if(functional)
+		functional = FALSE
+		for(var/atom/movable/AM in src)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:111]: Actually this should probably just be
`for (var/mob/living/carbon/corpse in src)`
Loops can do type checking themselves.

File: orbstation/modules/cargo/cadavers.dm [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+			if(istype(AM, /mob/living/carbon)) //unfreeze the organs of any contained carbon
+				var/mob/living/carbon/C = AM
+				for(var/part in C.internal_organs)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:114]: `for(var/obj/item/organ/part as anything in corpse.internal_organs)`

File: orbstation/modules/cargo/cadavers.dm [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+/datum/supply_pack/medical/cadaver/ethereal
+	name = "Medical Cadaver Crate (Ethereal)"
+	desc = "A preserved medical cadaver, for dissections, spare organs, or experiments. This one is an ethereal. For legal reason, the body's crystal core has been removed."
+	contains = list(/mob/living/carbon/human/species/ethereal)
+
+// Remove the crystal core, both so the ethereal doesn't regenerate and so the crew can't bulk-order crystal cores
+/datum/supply_pack/medical/cadaver/ethereal/generate()
+	. = ..()
+	var/mob/living/carbon/human/H = locate() in .
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:43]: Give this a real name

File: orbstation/modules/cargo/cadavers.dm [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	for (var/part in H.internal_organs)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:44]: `for (var/obj/item/organ/internal/heart/illegal_heart in corpse.internal_organs)`

File: orbstation/modules/cargo/cadavers.dm [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+		var/list/mob/dead/observer/candidates = poll_candidates_for_mob("Do you want to play as a vampire fugitive?", ROLE_FUGITIVE, null, 10 SECONDS, surprise_dracula, null)
+		if(LAZYLEN(candidates))
+			var/mob/dead/observer/chosen = pick(candidates)
+			surprise_dracula.key = chosen.key
+			surprise_dracula.mind.special_role = ROLE_FUGITIVE
+			to_chat(surprise_dracula, span_warning("You are a vampire, freshly escaped from imprisonment by NanoTrasen's secret Occultism Department. You've managed to stow away in place of a medical cadaver, bound for the unknown reaches of the Spinward Sector. \
+									You must do whatever it takes to evade capture."))
+			surprise_dracula.log_message("was made into a vampire fugitive by medical cadaver crate", LOG_GAME)
+			return
+	for (var/part in H.internal_organs) //each cadaver comes with a complimentary set of organs, held in stasis
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:31]: `for(var/obj/item/organ/part as anything in corpse.internal_organs)`
(assuming you rename H to corpse)

File: orbstation/modules/cargo/cadavers.dm [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+	///Whether a vampire has arrived on the station this round
+	var/vampire_arrival = FALSE
+	///Percent chance for a medical cadaver to be replaced with a vampire
+	var/vampire_chance = 1
+
+/datum/supply_pack/medical/cadaver/generate()
+	. = ..()
+	var/mob/living/carbon/human/H = locate() in .
+	H.death()
+	if(prob(vampire_chance) && !vampire_arrival)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:17]: You could and probably should invert this too.
```
. = ..()
var/mob/living/carbon/human/corpse = locate() in .
if (vampire_arrival || !prob(vampire_chance))
   corpse.death()
   return

... all the vampire stuff here
```

File: orbstation/modules/cargo/cadavers.dm [hunk 10]
```diff
@@ -0,0 +1,128 @@
+// Cadaver crates: crates that each contain one corpse, for "medical purposes".
+/datum/supply_pack/medical/cadaver
+	cost = CARGO_CRATE_VALUE * 5
+	access_view = ACCESS_MEDICAL
+	crate_type = /obj/structure/closet/crate/coffin/stasis
+	crate_name = "medical cadaver crate"
+
+	///Whether a vampire has arrived on the station this round
+	var/vampire_arrival = FALSE
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:9]: Would maybe be clearer as `vampire_arrived`

File: orbstation/modules/cargo/cadavers.dm [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+				var/mob/living/carbon/C = AM
+				for(var/part in C.internal_organs)
+					var/obj/item/organ/O = part
+					O.organ_flags &= ~ORGAN_FROZEN
+		//change icon state
+	return ..()
+
+/obj/structure/closet/crate/coffin/stasis/close(mob/living/user)
+	. = ..()
+	if(functional)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:122]:
```
if (!functional)
   return
```

File: orbstation/modules/cargo/cadavers.dm [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+		for(var/atom/movable/AM in src)
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:123]: You've got the idea by now probably

File: orbstation/modules/cargo/cadavers.dm (hunk 1 above)
Comment (lizardqueenlexi) [reply orbstation/modules/cargo/cadavers.dm]: Shit.

Changed this to only remove ethereal hearts in particular.

File: orbstation/modules/cargo/cadavers.dm [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	max_integrity = 70
+	material_drop = /obj/item/stack/sheet/iron
+	material_drop_amount = 5
+	open_sound = 'sound/machines/closet_open.ogg'
+	close_sound = 'sound/machines/closet_close.ogg'
+	open_sound_volume = 25
+	close_sound_volume = 50
+	can_install_electronics = FALSE
+	mob_storage_capacity = 1 //no mass stasis, sorry
+	var/functional = TRUE //whether the coffin still works or not
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:101]: Not a big deal but this being above as a triple slashed comment would be preferable (like `vampire_arrived` and `vampire_chance`), that format makes it show up in vscode intellisense when you hover over the variable.

File: orbstation/modules/cargo/cadavers.dm [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+	for(var/mob/living/carbon/corpse in src) //unfreeze the organs of any contained carbon
+		for(var/obj/item/organ/part as anything in corpse.internal_organs)
+			part.organ_flags &= ~ORGAN_FROZEN
+	return ..()
+
+/obj/structure/closet/crate/coffin/stasis/close(mob/living/user)
+	. = ..()
+	if(!functional)
+		return
+	for(var/mob/living/carbon/corpse in src) //freeze the organs of any contained carbon
```
Comment (Jacquerel) [inline orbstation/modules/cargo/cadavers.dm:123]: Is this ever called or required?
We set `functional` to `FALSE` when it is opened and do the stasis application manually when anything is placed into the crate.
When is it possible to close a still-functioning stasis coffin?

Comment (Jacquerel) [reply orbstation/modules/cargo/cadavers.dm:123]: If this never actually executes we might not need the `functional` variable at all.

Comment (lizardqueenlexi) [reply orbstation/modules/cargo/cadavers.dm]: Oops. This was an artifact from an older version of this code where this made more sense. Removed.

**Review by Jacquerel — APPROVED** [review]

**Pull request merged**